// @mui/x-date-pickers-pro/legacy/internal/hooks/validation/useDateRangeValidation.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
var _excluded = ["shouldDisableDate"];
import { useValidation, validateDate } from '@mui/x-date-pickers/internals';
import { isRangeValid, parseRangeInputValue } from '../../utils/date-utils';
export var validateDateRange = function validateDateRange(_ref) {
  var props = _ref.props,
      value = _ref.value,
      adapter = _ref.adapter;

  var _value = _slicedToArray(value, 2),
      start = _value[0],
      end = _value[1]; // for partial input


  if (start === null || end === null) {
    return [null, null];
  }

  var _shouldDisableDate = props.shouldDisableDate,
      otherProps = _objectWithoutProperties(props, _excluded);

  var dateValidations = [validateDate({
    adapter: adapter,
    value: start,
    props: _extends({}, otherProps, {
      shouldDisableDate: function shouldDisableDate(day) {
        return !!(_shouldDisableDate != null && _shouldDisableDate(day, 'start'));
      }
    })
  }), validateDate({
    adapter: adapter,
    value: end,
    props: _extends({}, otherProps, {
      shouldDisableDate: function shouldDisableDate(day) {
        return !!(_shouldDisableDate != null && _shouldDisableDate(day, 'end'));
      }
    })
  })];

  if (dateValidations[0] || dateValidations[1]) {
    return dateValidations;
  }

  if (!isRangeValid(adapter.utils, parseRangeInputValue(adapter.utils, value))) {
    return ['invalidRange', 'invalidRange'];
  }

  return [null, null];
};
export var isSameDateRangeError = function isSameDateRangeError(a, b) {
  return b !== null && a[1] === b[1] && a[0] === b[0];
};
export var useDateRangeValidation = function useDateRangeValidation(props) {
  return useValidation(props, validateDateRange, isSameDateRangeError);
};
