// @mui/x-date-pickers-pro/legacy/MultiInputDateRangeField/MultiInputDateRangeField.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import Stack from '@mui/material/Stack';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';
import { useMultiInputDateRangeField } from './useMultiInputDateRangeField';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
export var MultiInputDateRangeField = /*#__PURE__*/React.forwardRef(function MultiInputDateRangeField(inProps, ref) {
  var _useMultiInputDateRan = useMultiInputDateRangeField(inProps),
      startDate = _useMultiInputDateRan.startDate,
      endDate = _useMultiInputDateRan.endDate;

  return /*#__PURE__*/_jsxs(Stack, {
    spacing: 2,
    direction: "row",
    alignItems: "center",
    ref: ref,
    children: [/*#__PURE__*/_jsx(TextField, _extends({
      inputRef: startDate.inputRef
    }, startDate.inputProps)), /*#__PURE__*/_jsx(Typography, {
      children: "to"
    }), /*#__PURE__*/_jsx(TextField, _extends({
      inputRef: endDate.inputRef
    }, endDate.inputProps))]
  });
});
